dinstaller-lib/src/profile.rs

```diff
@@ -0,0 +1,125 @@
+use crate::manager::ManagerClient;
+use curl::easy::Easy;
+use jsonschema::JSONSchema;
+use serde_json;
+use std::{
+    error::Error,
+    fs, io,
+    io::{stdout, Write},
+    path::Path,
+    process::Command,
+};
+use tempfile::tempdir;
+
+/// Downloads a file a writes it to the stdout()
+///
+/// TODO: move this code to a struct
+/// TODO: add support for YaST-specific URLs
+/// TODO: do not write to stdout, but to something implementing the Write trait
+pub fn download(url: &str) -> Result<(), Box<dyn Error>> {
+    let mut easy = Easy::new();
+    easy.url(url)?;
+    easy.write_function(|data| {
+        stdout().write_all(data).unwrap();
+        Ok(data.len())
+    })?;
+    easy.perform()?;
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum ValidationResult {
+    Valid,
+    NotValid(Vec<String>),
+}
+
+/// Checks whether an autoinstallation profile is valid
+///
+/// ```
+/// # use dinstaller_lib::profile::{ProfileValidator, ValidationResult};
+/// # use std::path::Path;
+/// let validator = ProfileValidator::default_schema()
+///   .expect("the default validtor");
+///
+/// // you can validate a &str
+/// let wrong_profile = r#"
+///   { "product": { "name": "Tumbleweed" } }
+/// "#;
+/// let result = validator.validate_str(&wrong_profile).unwrap();
+/// assert!(matches!(ValidationResult::NotValid, result));
+///
+/// // or a file
+/// validator.validate_file(Path::new("share/examples/profile.json"));
+/// assert!(matches!(ValidationResult::Valid, result));
+/// ```
+pub struct ProfileValidator {
+    schema: JSONSchema,
+}
+
+impl ProfileValidator {
+    pub fn default_schema() -> Result<Self, Box<dyn Error>> {
+        let relative_path = Path::new("dinstaller-lib/share/profile.schema.json");
+        let path = if relative_path.exists() {
+            relative_path
+        } else {
+            Path::new("/usr/share/dinstaller-rs/profile.schema.json")
+        };
+        Self::new(path)
+    }
+
+    pub fn new(schema_path: &Path) -> Result<Self, Box<dyn Error>> {
+        let contents = fs::read_to_string(schema_path)?;
+        let schema = serde_json::from_str(&contents)?;
+        let schema = JSONSchema::compile(&schema).expect("A valid schema");
+        Ok(Self { schema })
+    }
+
+    pub fn validate_file(&self, profile_path: &Path) -> Result<ValidationResult, Box<dyn Error>> {
+        let contents = fs::read_to_string(profile_path)?;
```

Comment:
I expect that for error reporting case we would need to distinguish few cases like issue with our files like invalid syntax, wrong data and so on and user issues like file not provided. IO errors and so on and properly report it. Challenging part for me is how to make it not hardcoded in lib and instead provide some nice errors, that cli can meaningfully print to user.

Reply:
Yes, for me I/O problems here are encoded with the `Result::Err` variant, but a validation problem (an invalid profile) is just a result (encoded as `Ok<ValidationResult>`.

Reply:
well, here I mean we should distringuish errors from files we maintain like schema and errors from provided json file like disrupted load and corrupted file and so on. Not exactly validation error.